libs/runtime/ebpf_ring_buffer.c

```diff
@@ -8,78 +8,298 @@
 }
 
-inline static _Ret_notnull_ ebpf_ring_buffer_record_t*
-_ring_record_at_offset(_In_ const ebpf_ring_buffer_t* ring, size_t offset)
+/**
+ * @brief Get the producer reserve offset.
+ *
+ * Uses read-acquire to ensure producers see the latest value before trying compare-exchange.
+ *
+ * @param[in] ring Pointer to the ring.
+ * @return Producer reserve offset.
+ */
+inline static size_t
+_ring_read_producer_reserve_offset_acquire(_In_ const ebpf_ring_buffer_t* ring)
 {
-    return (ebpf_ring_buffer_record_t*)&ring->shared_buffer[offset % ring->length];
+    return ReadULong64Acquire(&ring->producer_reserve_offset);
 }
 
+/**
+ * @brief Atomically advance the producer reserve offset to reserve space for a new record.
+ *
+ * Reserves space between expected_value and new_value if expected_value is returned.
+ *
+ * @param[in] ring Pointer to the ring.
+ * @param[in] new_value New reserve offset to attempt to set.
+ * @param[in] expected_value Expected current reserve offset.
+ * @returns expected_value if the compare-exchange succeeded, otherwise the previous value.
+ */
+inline static size_t
+_ring_exchange_producer_reserve_offset(_Inout_ ebpf_ring_buffer_t* ring, size_t new_value, size_t expected_value)
+{
+    return (uint64_t)ebpf_interlocked_compare_exchange_int64(
+        (volatile int64_t*)&ring->producer_reserve_offset, new_value, expected_value);
+}
+
+/**
+ * @brief Get record size from data size (includes header).
+ *
+ * Adds 8 bytes for the header and pads to 8 byte alignment.
+ *
+ * @param[in] data_length Length of the data in the record.
+ * @return Size of the record.
+ */
+inline static size_t
+_ring_record_size(size_t data_length)
+{
+    return ((EBPF_OFFSET_OF(ebpf_ring_buffer_record_t, data) + data_length) + 7) & ~7;
+}
+
+/**
+ * @brief Get the length of the record from the header.length.
+ *
+ * Excludes the lock and discard bits.
+ *
+ * @param[in] header_length Record header.length.
+ * @return Length of the record.
+ */
+inline static uint32_t
+_ring_header_length(_In_ uint32_t header_length)
+{
+    return header_length & ~(EBPF_RINGBUF_LOCK_BIT | EBPF_RINGBUF_DISCARD_BIT);
+}
+
+/**
+ * @brief Check if the record is locked from the header.
+ *
+ * @param[in] header_length Record header.length.
+ * @retval true The record is locked.
+ * @retval false The record is not locked.
+ */
+inline static bool
+_ring_header_locked(_In_ uint32_t header_length)
+{
+    return (header_length & EBPF_RINGBUF_LOCK_BIT) != 0;
+}
+
+/**
+ * @brief Check if the record is discarded from the header.
+ *
+ * @param[in] header_length Record header.length.
+ * @retval true The record is discarded.
+ * @retval false The record is not discarded.
+ */
+inline static bool
+_ring_header_discarded(_In_ uint32_t header_length)
+{
+    return (header_length & EBPF_RINGBUF_DISCARD_BIT) != 0;
+}
+
+/**
+ * @brief Get the length of the ring.
+ *
+ * @param[in] ring Pointer to the ring.
+ * @return Length of the record.
```

Comment:
240 says "length of the ring" but 243 says "length of the record".
These sound contradictory to me.